packages/lesswrong/server/languageModels/autoTagCallbacks.ts

```diff
@@ -0,0 +1,246 @@
+ *        openai api fine_tunes.list
+ *    This will output a list of fine-tuned models you've created. The field
+ *    you want is named `fine_tuned_model` and looks like
+ *        babbage:ft-personal-YYYY-MM-DD-HH-mm-ss
+ *    For each tag, find the corresponding model ID, go to the corresponding tag
+ *    page, and put the model ID in the "auto-tag classifier model ID" field.
+ *
+ * 10. Generate a comparison list between human-applied and auto-applied tags for
+ *     the test set.
+ *        scripts/serverShellCommand.sh 'Globals.evaluateTagModels("ml/tagClassificationPostIds.test.json", "ml/tagClassificationTestSetResults.txt")'
+ *     This produces a text file ml/tagClassificationTestSetResults.txt with a
+ *     list of post titles/links, how humans tagged them, and how the trained
+ *     models tagged them. Make sure this looks reasonable.
+ */
+
+const bodyWordCountLimit = 1500;
+const tagBotAccountSlug = new DatabaseServerSetting<string|null>('languageModels.autoTagging.taggerAccountSlug', null);
+
+
+/**
+ * Preprocess HTML before converting to markdown to be then converted into a
+ * language model prompt. Strips links, and replaces images with their alt text
+ * or with "IMAGE". We do this to prevent URLs (which tend to be long, and
+ * uninformative, and prone to future distribution shifts if we change how our
+ * image hosting works) from chewing up limited context-window space.
+ *
+ * TODO: Replace images with their alt text
+ * TODO: Replace any Unicode characters that are going to cause trouble for the GPT-3 tokenizer
+ */
+function preprocessHtml(html: string): string {
+  const $ = cheerioParse(html) as any;
+  $('a').contents().unwrap();
+  return $.html();
+}
+
+export async function postToPrompt({template, post, promptSuffix, postBodyCache}: {
+  template: LanguageModelTemplate,
+  post: DbPost,
+  promptSuffix: string
+  // Optional mapping from post ID to markdown body, to avoid redoing the html-to-markdown conversions
+  postBodyCache?: PostBodyCache
+}): Promise<string> {
+  const {header, body} = template;
+  
+  const markdownPostBody = postBodyCache?.preprocessedBody?.[post._id] ?? preprocessPostBody(post);
+  
+  const linkpostMeta = post.url ? `\nThis is a linkpost for ${post.url}` : '';
+  
+  return substituteIntoTemplate({
+    template,
+    maxLengthTokens: parseInt(header["max-length-tokens"]),
+    truncatableVariable: "text",
+    variables: {
+      title: post.title,
+      linkpostMeta,
+      text: markdownPostBody,
+      tagPrompt: promptSuffix,
+    }
+  });
+}
+
+function preprocessPostBody(post: DbPost): string {
+  const postHtml = post.contents?.html;
+  const markdownPostBody = postHtml ? dataToMarkdown(preprocessHtml(postHtml), "html") : "";
+  return markdownPostBody;
+}
+
+export type PostBodyCache = {preprocessedBody: Record<string,string>}
+export function generatePostBodyCache(posts: DbPost[]): PostBodyCache {
+  const result = {preprocessedBody: {}};
```

Comment:
Same as https://github.com/ForumMagnum/ForumMagnum/pull/6346/files#r1057331655, though I think something like:
```
const preprocessedBody: Record<string, string> = {};
for (...) {
  preprocessedBody[post._id] = ...;
}
return { preprocessedBody };
```

Reply:
That link doesn't go anywhere useful and I'm not sure what you mean